designs/sam-app-config.md

```diff
@@ -0,0 +1,201 @@
+SAM CLI App Level Config
+====================================
+
+
+What is the problem?
+--------------------
+
+Today users of SAM CLI need to invoke the CLI directly with all parameters supplied to its commands.
+
+for e.g: `sam build --use-container --debug`
+
+But often, during the lifecycle of building and deploying a serverless application. the same commands get run repeatedly to build, package and deploy, before solidifying into the final application. 
+
+These CLI commands are often long and have many changing parts.
+
+Have a look at the following series of workflows
+
+
+* `sam build --use-container --template ... --parameter-overrides=... --skip-pull-image --manifest ...`
+
+* `sam package --s3-bucket ... --template-file ... --output-template-file ... --s3-prefix ... --kms-key-id ...`
+
+* `sam deploy deploy --template-file ... --stack-name ... --capabilities ... --tags ... --parameter-overrides ... --kms-key-id ...`
+
+If this could be condensed into a series of workflows that look like
+
+* `sam build`
+* `sam package`
+* `sam deploy`
+
+That would be a huge user experience win.
+
+What will be changed?
+---------------------
+
+The suite of commands supported by SAM CLI would be aided by looking for a configuration file thats locally located to the template.yaml by default. 
+
+`.aws-sam/sam-app-config`
```

Comment:
Why `sam-app-config`? Should we simplify this further to samconfig? 

I am confused on the placement within `.aws-sam`. To this point, this directory is generated by the cli directly (to ensure correct permissions, mainly). With this, we are suggestion to customers that they need to now create this directory. This might be ok if the entry point was through a command (sam config generate) but we aren't that far yet. Why not place this within the root of the project?

Reply:
We can move it up to the project root, to me this configuration was very specific to the app the user was working on at that point. Also might prevent accidental checking it in to source control, since we ignore the .aws-sam directory.

Reply:
`sam-app-config` because it doesn't change the inherent behavior of SAM CLI, its a pass through of parameters.